### DISCUSSION TITLE
Best pattern for executing sql right after a Session is committed?

### DISCUSSION THREAD

**Discussion by jfly** [sqlalchemy/sqlalchemy#9116] [Usage Questions] [answered]
What's the best pattern for executing some sql right after a Session is committed?

I've tried registering an `after_commit` event listener, but I get an `InvalidRequestError: A transaction is already begun on this Session` if I try to use the session (I know the underlying database transaction is done and committed, but sqlalchemy's bookkeeping considers the Session to not yet be ready for reuse).

Something that does work is to create a brand new Session. That seems to work fine, but I'd love understand if there are any downsides to that approach. Perhaps it means that we're creating 2 connections to the database when we really only need to use 1 serially?

## Demonstration of the two approaches

First, start a database:

```
$ docker run -e MYSQL_ALLOW_EMPTY_PASSWORD=yes -e MYSQL_DATABASE=test -p 3306:3306 mysql:8
```

Second, create a `postcommit.py` file like this:

```python
#!/usr/bin/env python

import sys

from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import text
from sqlalchemy.orm import sessionmaker

mode = sys.argv[1]

db_engine = create_engine("mysql+pymysql://root:@localhost:3306/test")
Session = sessionmaker(bind=db_engine, autoflush=True, autocommit=False, future=True)

session = Session()

def postcommit(postcommit_session):
    assert postcommit_session is session

    if mode == 'current-session':
        with postcommit_session.begin():
            result = postcommit_session.execute(
                text("SELECT 'query from postcommit';")
            ).one()
            print(f"result from postcommit: {result}")
    elif mode == 'new-session':
        new_session = Session()
        with new_session.begin():
            result = new_session.execute(text("SELECT 'query from postcommit';")).one()
            print(f"result from postcommit: {result}")
    else:
        print(f"Unrecognized mode: {mode}")


def do_transaction():
    with session.begin():
        event.listen(session, "after_commit", postcommit, once=True)
        result = session.execute(text("SELECT 'query from do_transaction';")).one()
        print(f"result from do_transaction: {result}")


do_transaction()
```

### Note how things crash if you try to reuse the current session

```
$ ./postcommit.py current-session
result from do_transaction: ('query from do_transaction',)
Traceback (most recent call last):
  File "./postcommit.py", line 53, in <module>
    do_transaction()
  File "./postcommit.py", line 50, in do_transaction
    print(f"result from do_transaction: {result}")
  File "/home/jeremy/src/github.com/joinhonor/h4-thrift-server/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/engine/util.py", line 237, in __exit__
    self.rollback()
  File "/home/jeremy/src/github.com/joinhonor/h4-thrift-server/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/jeremy/src/github.com/joinhonor/h4-thrift-server/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/jeremy/src/github.com/joinhonor/h4-thrift-server/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/engine/util.py", line 233, in __exit__
    self.commit()
  File "/home/jeremy/src/github.com/joinhonor/h4-thrift-server/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 839, in commit
    self.session.dispatch.after_commit(self.session)
  File "/home/jeremy/src/github.com/joinhonor/h4-thrift-server/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 334, in __call__
    fn(*args, **kw)
  File "/home/jeremy/src/github.com/joinhonor/h4-thrift-server/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 1695, in go
    return once_fn(*arg, **kw)
  File "./postcommit.py", line 34, in postcommit
    with session.begin():
  File "<string>", line 2, in begin
  File "/home/jeremy/src/github.com/joinhonor/h4-thrift-server/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 375, in warned
    return fn(*args, **kwargs)
  File "/home/jeremy/src/github.com/joinhonor/h4-thrift-server/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1331, in begin
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.
```

### But things work if you create a brand new session

```
$ ./postcommit.py new-session
result from do_transaction: ('query from do_transaction',)
result from postcommit: ('query from postcommit',)
```

**Comment by zzzeek** [comment, marked as answer]
how come you want to automate the emitting of SQL *after* the actual COMMIT call? Why not before?  much easier to do it that way.  What state do you expect in the database after COMMIT that's not there right before it's called?

**Reply by jfly** [reply]
@zzzeek, we're implementing the [transactional outbox pattern](https://microservices.io/patterns/data/transactional-outbox.html) to ensure that we publish events if and only if a transaction gets committed. We also want to try to deliver those events as quickly as possible, so we'd like to (try to) process them immediately after commit rather than waiting for our polling message relay to process them.

Consider this sequence of events on a hypothetical users api service:

1. api call starts
2. begin a transaction
3. insert a row into the users table
4. record an entry in the outbox table to remember to emit an event for that user getting created
5. commit transaction
6. (in postcommit):
   - actually emit the event (this looks like some api call to something like kafka)
   - begin new transaction, delete that row from the outbox table, commit transaction
7. api call ends

It would sure be nice if we could use the same Session throughout this, because we're actually using sqlalchemy scoped sessions to share a single session throughout our entire request.

**Reply by zzzeek** [reply]
OK well, within after_commit(), that's within the Session's process of completing the transaction, so if you want to run SQL also in there using ORM patterns, you would make a separate Session in that event hook with which to run ORM queries.  there's no disadvantage to this since the Session you're inside of has closed out it's connection state in any case.

**Reply by jfly** [reply]
Great, thanks @zzzeek! Also good to know that we're not going to end up creating more db connections than necessary.

I'm sure this is a long shot, but: is there any chance sqlalchemy would be open to adding a new event type specifically for this use case? It's easy enough to create a new session in the simple code example I shared above, but in practice, we're using scoped sessions tied to a flask request, and it would sure be cleaner if we could just keep using that same Session.

**Reply by zzzeek** [reply]
event hooks are used to be able to run code within parts of SQLAlchemy that would normally not be possible - hence all the hooks are "within" an existing process.

Here, you are looking for something you want to do after your program actually calls .commit().  There's not an actual need for an event hook within SQLAlchemy for that, your own code can provide its own method that calls session.commit() then run_my_events() directly after that.

That said, you might have better luck using the [after_transaction_end](https://docs.sqlalchemy.org/en/14/orm/events.html#sqlalchemy.orm.SessionEvents.after_transaction_end) event.  this is the event that zope.transaction hooks into for its sqlalchemy extension and integrates with similar behavior as you describe, where things like sending emails etc. after commit has succeeded occur.   but even then, by requesting to invoke new SQL within the event, it means you are nesting the "begin a new transaction" process within the "end the previous transaction" process and this is inherently awkward.

**Reply by jfly** [reply]
That makes a lot of sense. You're totally right that an individual application using sqlalchemy can structure its code in a way such that sqlalchemy doesn't need to provide such a hook.

However (and I get that I'm being a bit hyperbolic here), one could make a case that not having a "pure sqlalchemy" way to do this does limit the potential richness of the sqlalchemy ecosystem as a whole. For example, I'm on an internal infra team, and we're trying to build a reusable "transactional outbox" library for our individual services to use, and it would be nice if our library worked just fine with a sqlalchemy session. Instead, right now we need something else: either some postcommit callback or some mechanism to truly create a new session (not a scoped session). Every conversation I've had internally with folks about this leaves them surprised that this is necessary, and concerned that we're misusing or misunderstanding sqlalchemy. `zope.transaction` is a neat example, I'll have to check it out.

I tried out `after_transaction_end`, and it seems to half-work. Demonstration:

<details>
<summary>Slightly modified postcommit.py from above</summary>

```
#!/usr/bin/env python

import sys

from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import text
from sqlalchemy.orm import sessionmaker

mode = sys.argv[1]

db_engine = create_engine("mysql+pymysql://root:@localhost:3306/test")
Session = sessionmaker(bind=db_engine, autoflush=True, autocommit=False, future=True)


def postcommit(postcommit_session, transaction):
    if transaction.parent is not None:
        return

    with postcommit_session.begin():
        result = postcommit_session.execute(
            text("SELECT 'query from postcommit';")
        ).one()
        print(f"result from postcommit: {result}")


def do_transaction():
    if mode == 'ctx-manager':
        session = Session()
        with session.begin():
            event.listen(session, "after_transaction_end", postcommit, once=True)
            result = session.execute(text("SELECT 'query from do_transaction';")).one()
            print(f"result from do_transaction: {result}")
    elif mode == "try-except":
        session = Session()
        try:
            event.listen(session, "after_transaction_end", postcommit, once=True)
            result = session.execute(text("SELECT 'query from do_transaction';")).one()
            print(f"result from do_transaction: {result}")
            session.commit()
        except:
            session.rollback()
            raise
    else:
        print(f"Unrecognized mode: {mode}")


do_transaction()
```

</details>

## This works great if we're managing the session state with try...except

```
$ ./postcommit.py try-except
result from do_transaction: ('query from do_transaction',)
result from postcommit: ('query from postcommit',)
```

## This crashes if we're using the `session.begin` context manager

```
$ ./postcommit.py ctx-manager
result from do_transaction: ('query from do_transaction',)
Traceback (most recent call last):
  File "./postcommit.py", line 48, in <module>
    do_transaction()
  File "./postcommit.py", line 33, in do_transaction
    print(f"result from do_transaction: {result}")
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/engine/util.py", line 237, in __exit__
    self.rollback()
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 208, in raise_
    raise exception
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/engine/util.py", line 233, in __exit__
    self.commit()
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 843, in commit
    self.close()
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 928, in close
    self.session.dispatch.after_transaction_end(self.session, self)
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 334, in __call__
    fn(*args, **kw)
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 1695, in go
    return once_fn(*arg, **kw)
  File "./postcommit.py", line 20, in postcommit
    with postcommit_session.begin():
  File "<string>", line 2, in begin
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 309, in warned
    return fn(*args, **kwargs)
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1320, in begin
    if self._autobegin():
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1260, in _autobegin
    trans = SessionTransaction(self, autobegin=True)
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 527, in __init__
    TransactionalContext._trans_ctx_check(session)
  File "/home/jeremy/src/github.com/joinhonor/auth-api/.direnv/python-3.8.10/lib/python3.8/site-packages/sqlalchemy/engine/util.py", line 199, in _trans_ctx_check
    raise exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Can't operate on closed transaction inside context manager.  Please complete the context manager before emitting further commands.
```

Is this a bug? Would it be behavior you'd be open to changing in sqlalchemy?

**Reply by zzzeek** [reply]
it's really not a bug, no.  these are hooks inside the internals of the Session, so state changes to the Session while it's in the process of changing its state is never going to be non-awkward and non-wrong-feeling.

this is the thing - the Session does not emit SQL *without* a "transaction" , where by transaction we mean Session's client-representation of a "transaction" - whether there is SQL connectivity inside that "transaction" or not is optional, but the moment you emit SQL in that "transaction", connectivity + real DB transaction is also pulled into that.   So I dont really know what a model for "we want to start a new transaction while the old one is still finishing up and running some SQL" would be.

this is the reason this points to an external framework being more appropriate for "commit this database transaction, then start a new transaction, get some data, send an email, then rollback or commit *that* transaction.      The tendency here is people want to turn SQLAlchemy into a framework and we are not in the framework business.   

I think in your ultimate use case, this additional data you want to get at transaction end you would prefer not to have to "check out a brand new connection from the pool, start transaction, run SELECT, rollback transaction, return connection to pool" - you'd want to use the connection that's already available before the transaction ends.   Here's the best (and IMO pretty easy) way to do that -  do the SQL part of things in `before_commit()` and the "send an email" things in `after_commit()` or `after_transaction_end()`.  You can share the state between the two events using [session.info](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.info), which is there for these kinds of things.

That way you aren't unnecessarily making a whole new connection pool checkout + DB transaction.

**Reply by jfly** [reply]
> do the SQL part of things in before_commit() and the "send an email" things in after_commit() or after_transaction_end().

We truly _do_ want to do sql things in the `after_{commit,transaction_end}` callback, though. After sending the email (or whatever row we're processing in the transactional outbox), we need to delete that row/mark it processed so our "message relay" doesn't re-do that work.

That all said, I understand if this problem should be considered outside of sqlalchemy's purview, in favor of some "framework" layer on top. We can definitely build something simple ourselves, but do you have any recommendations of such frameworks to investigate?

Thanks again, this has been really useful information.

**Reply by zzzeek** [reply]
> > do the SQL part of things in before_commit() and the "send an email" things in after_commit() or after_transaction_end().
> 
> We truly _do_ want to do sql things in the `after_{commit,transaction_end}` callback, though. After sending the email (or whatever row we're processing in the transactional outbox), we need to delete that row/mark it processed so our "message relay" doesn't re-do that work.

OK so again what's the problem using a distinct Session()?  Using a separate session for "out of band" transactions like things for logging and such is a pretty standard approach.

**Reply by jfly** [reply]
There's no problem using a distinct Session(), it just triggered a lot of internal discussion about if it's really necessary, or a sign that we're mis-using sqlalchemy. I now know the answer to that question!

> Using a separate session for "out of band" transactions like things for logging and such is a pretty standard approach.

That's interesting! Are there any examples/docs you can link to about this? Would love to learn more.

**Reply by zzzeek** [reply]
it's been awhile since I did this on the job but you would have a second transaction for cases like an application that writes its logs to the database, or has an audit trail or live-updating status notifier that uses the database as the backing storage.    So if you had a large ETL job that nonetheless needs live status updates, those have to be committed for a front end to be able to consume those messages as they happen, even if the ETL job itself is still ongoing within a larger transaction.    Similarly, reporting problems and such that occur when the main transaction is rolled back; when you're in the main rollback, it's nice to have a totally different transaction that you open on an ad-hoc basis in order to update status.

The secondary session can also be using a lower isolation level or even autocommit isolation to keep its overhead / contention with other processes lower.

**Reply by jfly** [reply]
Ooh those are two interesting examples I have no experience with, but both make a lot of sense.

Thanks again for being so responsive and informative in your answers here!
